
# Coding in Class
Hands-on modules to highlight what you have read, but also to set you up for the HW -> many things are similar 


While I go through, I have written errors in my code

*   When errors come I will ask to see if anyone can spot the problem
*   Why? Errors happen. It is good for us to work through bugs in our code by breaking things down

# Lesson #1 -- PY4E CH 1 and 2

At the most basic levels of coding, code allows us to interface with the computer to get an output.



In [ ]:
print('Hello World')

Hello World


We can also assign data to variables or place holders. 

In [ ]:
w = 8					
print(w) 			

	# placeholders can be anything, but it is best practice to keep them descriptive 


8


In [ ]:
# We can also use code to manipulate these variables. 
w = 8					
print(w * 14)


# while we have used the parameter 'w' in the expression, the value of 'w' stays the same and we haven't 'saved' or expression

112


In [ ]:
# We can overwrite the variable 'w' by assigning the expression again to 'w'
w = 8					
print(w)
w = w * 14				
print(w) 	

# In this simple code we also make a fundamental point: logical chain

8
112


In [ ]:
# But it is best practice to save our expression by creating a new parameter 'y' incase we want to call 'w' in the future 
w = 8					
print(w)
y = w * 14				
print(y) 	

# In this simple code we also make a fundamental point: logical chain

8
112


While this is a basic illustration of how we can assign data to variables and interact with those variables, let’s turn to writing some basic code that we can interact with. 

In [ ]:
height = input("How tall are you (round to the nearest foot)?")
print(height) 


How tall are you (round to the nearest foot)?6
6


Let's go back and make this a bit more interesting.

In [ ]:
height = input("How tall are you (round to the nearest foot)?")
print("you are about " height " feet tall")  #Error


SyntaxError: ignored

In [ ]:
height = input("How tall are you (round to the nearest foot)?")
print("you are about " , height , " feet tall")  #individual data --> (string, variable, string)


How tall are you (round to the nearest foot)?6
you are about  6  feet tall


Say we want to report the height of our respondent in inches.

In [ ]:
height = input("How tall are you (round to the nearest foot)?")
print("you are about " + height +" feet tall")
h_inch = height * 12
print("you are about " ,  h_inch ," inches tall")

How tall are you (round to the nearest foot)?6
you are about 6 feet tall
you are about  666666666666  inches tall


Why we are getting 666666666666 ?		

While height is referencing the input variable that variable is a string variable. Python is manipulating what you are giving it. It has multiplied the string ‘6’ twelve times. 

Python sees all inputted data as string data. 


In [ ]:
height = input("How tall are you (round to the nearest foot)?")
print("you are about " , height ," feet tall")
h = int(height)                                         #Must identify the data as an integer
h_inch = height * 12
print("you are about " , h_inch ," inches tall")


How tall are you (round to the nearest foot)? 6
you are about   6  feet tall
you are about   6 6 6 6 6 6 6 6 6 6 6 6  inches tall


Still a problem.

Violation of the logical chain. We change height to an integer and assign it to h, but we do not use h in the next line of code.

In [ ]:
height = input("How tall are you (round to the nearest foot)?")
print("you are about " , height ," feet tall")
h = int(height)
h_inch = h * 12                                          # maintain the logical chain
print("you are about " , h_inch ," inches tall")

How tall are you (round to the nearest foot)?6
you are about  6  feet tall
you are about  72  inches tall


# Using Command Line

While we will not be using command line in this class, it is a good tool to know of and be confortable with. The command line represents the path your computer uses to identify and open files. By understanding this pathway process, you will have a better sense of how to access and open files through python and in the future in R.

First, we will create a folder on our Desktop called 'PAI_724'. Next we will save the last 'height function' we coded as a new colab file named 'Test_1' (i.e., go to 'File' > 'New notebook') and download it as a *.py* file. 

When downloading, Google will prompt you that the 'file might harm your computer'. This is because .py files and python files more broadly are executing a function on your computer, and google is not checking the safety of this function. Generally you should know what you are downloading -- because we wrote this function ourselves, we know the function is safe. Thus, please select 'Keep'. 

Once it is downloaded, place the file in the PAI_724 file. 

Next, open 'Command Prompt' by seaching for it in your 'Start'. When opened, you will see somthing like the following: 

>*Microsoft Windows [Version 10.0.22000.856]*
>
>*(c) Microsoft Corporation. All rights reserved.*
>
>*C:\Users\ambro>*

Generally there are two commands we are interested in: (1) *dir*, which offers us the directory at the current position, and (2) *cd*, which advances the current directory.

At the moment, the directory is set at the largest level -- the computer. 

Let's see what is at this position using *dir*. Type *dir* and hint the 'enter' key.

>*Microsoft Windows [Version 10.0.22000.856]*
>
>*(c) Microsoft Corporation. All rights reserved.*

>*C:\Users\ambro>* **dir** 

Resulting in:

>*Directory of C:\Users\ambro*

>*09/05/2022  05:22 PM    {DIR}          Desktop*
>
>*08/29/2022  02:13 PM    {DIR}          Documents*
>
>*09/07/2022  12:34 PM    {DIR}          Downloads *

>*C:\Users\ambro>*

We can advance our directory into 'Desktop' using *cd*. Type *cd* 'Desktop'and hint the 'enter' key.

>*Directory of C:\Users\ambro*

>*09/05/2022  05:22 PM    {DIR}          Desktop*
>
>*08/29/2022  02:13 PM    {DIR}          Documents*
>
>*09/07/2022  12:34 PM    {DIR}          Downloads *

>*C:\Users\ambro>* **cd Desktop**

Resulting in the position of the command line being advanced into the 'Desktop' folder:

>*C:\Users\ambro\Desktop>*

Again, we can see what is in the directory in this new location by typing *dir*:

>*C:\Users\ambro\Desktop>* **dir**

Resulting in:

>*Directory of C:\Users\ambro\Desktop*

>*02/08/2022  12:12 PM    {DIR}          EJ Councils*
>
>*08/29/2022  02:38 PM   {DIR}          PAI 724*
>
>*08/15/2022  11:12 AM            23,530   selectionTable_ans2.png *

>*C:\Users\ambro\Desktop>*

Here, we now see the difference between folders (identified by {DIR}) and files. For files we see their size in bytes. Where we could use the *cd* function to progress into specific folders, we need specific commands to call files.  We will call a python file in our next step. 

First using *cd* and *dir*, move into the 'PAI 724' file and call its directory.

>*C:\Users\ambro\Desktop>* **cd PAI_724**

>*C:\Users\ambro\Desktop\PAI_724>* **dir**

Finally we can call out python file 'Test_1' using the *python* command

>*C:\Users\ambro\Desktop\PAI_724>* **python Test_1.py** 

After calling the python file, you should be able to execute the prompts as you did in Collaboratory.